keylime/tpm/tpm_main.py

```diff
@@ -1200,6 +1273,48 @@ def check_quote(
             hash_alg,
         )
 
+    def check_quote_timing(self, previous_clockinfo, quote):
+        # Sanity check quote clock information
+
+        current_clockinfo = {}
+
+        if previous_clockinfo:
+
+            pclki = TPMClockInfo.from_json(previous_clockinfo)
+
+            retout, success = self._tpm2_printquote(quote, False)
+            if not success:
+                return "tpm2_print failed with " + str(retout), current_clockinfo
+
+            tpm_data_str_dict = config.yaml_to_dict(retout, add_newlines=False, logger=logger)
+            if tpm_data_str_dict is None:
+                return "yaml output of tpm2_print could not be parsed!", current_clockinfo
+
+            cclki = TPMClockInfo.from_json(tpm_data_str_dict)
+
+            resetdiff = cclki.resetcount - pclki.resetcount
+            restartdiff = cclki.restartcount - pclki.restartcount
+
+            if resetdiff < 0:
+                return "resetCount value decreased on TPM between two consecutive quotes", current_clockinfo
+
+            if restartdiff < 0:
+                return "restartCount value decreased on TPM between two consecutive quotes", current_clockinfo
+
+            if cclki.safe != 1:
+                return "clock safe flag is disabled", current_clockinfo
+
+            if not (resetdiff and restartdiff):
+                if cclki.clock - pclki.clock <= 0:
+                    return (
+                        "clock timestamp did issued by TPM did not increase between two consecutive quotes",
+                        current_clockinfo,
+                    )
+
+            current_clockinfo = cclki.to_json()
```

Comment:
Why not return `TPMClockInfo`. I would to the json conversion as close as possible to the database storage.

Reply:
Cannot disagree, will change